core/emulate.c

```diff
@@ -0,0 +1,807 @@
+        break;
+    case 2:
+        op->mem.ea = insn_fetch_u16(ctxt);
+        break;
+    case 4:
+        op->mem.ea = insn_fetch_u32(ctxt);
+        break;
+    }
+}
+
+static void decode_op_imm(em_context_t *ctxt,
+                          em_operand_t *op)
+{
+    op->type = OP_IMM;
+    op->size = ctxt->operand_size;
+    switch (op->size) {
+    case 1:
+        op->value = insn_fetch_u8(ctxt);
+        break;
+    case 2:
+        op->value = insn_fetch_u16(ctxt);
+        break;
+    case 4:
+        op->value = insn_fetch_u32(ctxt);
+        break;
+    case 8:
+        op->value = insn_fetch_u32(ctxt);
+        break;
+    }
+}
+
+static void decode_op_simm8(em_context_t *ctxt,
+                            em_operand_t *op)
+{
+    op->type = OP_IMM;
+    op->size = 1;
+    op->value = (int64)((int8)insn_fetch_u8(ctxt));
+}
+
+static void decode_op_acc(em_context_t *ctxt,
+                          em_operand_t *op)
+{
+    op->type = OP_REG;
+    op->size = ctxt->operand_size;
+    op->reg.index = REG_RAX;
+}
+
+static void decode_op_di(em_context_t *ctxt,
+                         em_operand_t *op)
+{
+    op->type = OP_MEM;
+    op->size = ctxt->operand_size;
+    op->mem.ea = READ_GPR(REG_RDI, ctxt->address_size);
+    op->mem.seg = SEG_ES;
+}
+
+static void decode_op_si(em_context_t *ctxt,
+                         em_operand_t *op)
+{
+    op->type = OP_MEM;
+    op->size = ctxt->operand_size;
+    op->mem.ea = READ_GPR(REG_RSI, ctxt->address_size);
+    op->mem.seg = SEG_DS;
+    if (ctxt->override_segment) {
+        op->mem.seg = ctxt->override_segment;
+    }
+}
+
+/* Soft-emulation */
+static void em_mov(struct em_context_t *ctxt)
+{
+    memcpy(&ctxt->dst, &ctxt->src1, ctxt->operand_size);
+}
+
+static void em_movzx(struct em_context_t *ctxt)
+{
+    uint64_t value = 0;
+    memcpy(&value, &ctxt->src1, ctxt->operand_size);
+    ctxt->dst.value = value;
+}
+
+static void em_xchg(struct em_context_t *ctxt)
+{
+    uint64_t src1, src2;
+    src1 = ctxt->src1.value;
+    src2 = ctxt->src2.value;
+    ctxt->src1.value = src2;
+    ctxt->src2.value = src1;
+    operand_write(ctxt, &ctxt->src1);
+    operand_write(ctxt, &ctxt->src2);
+}
+
+em_status_t em_decode_insn(struct em_context_t *ctxt, uint8_t *insn)
+{
+    uint8_t b;
+    uint64_t flags;
+    struct em_opcode_t *opcode;
+    const struct em_opcode_t *opcode_group;
+
+    switch (ctxt->mode) {
+    case EM_MODE_PROT16:
+        ctxt->operand_size = 2;
+        ctxt->address_size = 2;
+        break;
+    case EM_MODE_PROT32:
+        ctxt->operand_size = 4;
+        ctxt->address_size = 4;
+        break;
+    case EM_MODE_PROT64:
+        ctxt->operand_size = 4;
+        ctxt->address_size = 8;
+        break;
+    default:
+        return EM_ERROR;
+    }
+    ctxt->override_segment = SEG_NONE;
+    ctxt->override_operand_size = 0;
+    ctxt->override_address_size = 0;
+    ctxt->insn = insn;
+    ctxt->lock = 0;
+    ctxt->rep = 0;
+    ctxt->len = 0;   
```

Comment:
nit: Avoid trailing whitespace. I also see this issue elsewhere in the file since my editor is set to highlight trailing whitespace. You may consider doing the same.